Mathlib/RingTheory/MvPolynomial/NewtonIdentities.lean

```diff
@@ -0,0 +1,327 @@
+    card t.fst = k := by
+  simp_rw [pairs, mem_filter, PairsPred] at ht
+  exact Or.resolve_right (le_iff_eq_or_lt.mp ht.2.1) hnlt
+
+def weight (k : ℕ) (t : Finset σ × σ) : MvPolynomial σ R :=
+  (-1) ^ (card t.fst) * ((∏ a in t.fst, X a) * (X t.snd) ^ (k - card t.fst) : MvPolynomial σ R)
+
+def T_map (t : Finset σ × σ) : Finset σ × σ :=
+  if t.snd ∈ t.fst then (t.fst.erase t.snd, t.snd) else (insert t.snd t.fst, t.snd)
+
+/-- Needed for `Finset.sum_involution` -/
+def T_map_restr (t : Finset σ × σ) (_ : t ∈ pairs σ k) := T_map σ t
+
+open Classical in
+theorem T_map_pair (t : Finset σ × σ) (h : t ∈ pairs σ k) : T_map_restr σ t h ∈ pairs σ k := by
+  rw [pairs, mem_filter, PairsPred] at *
+  simp_rw [T_map_restr, T_map]
+  split_ifs with h1
+  · simp_all
+    apply And.intro
+    · exact le_step h
+    · by_contra h2
+      have h3 : t.fst.Nonempty
+      · use t.snd
+        apply h1
+      rw [← h2] at h
+      exact not_le_of_lt (sub_lt (card_pos.mpr h3) zero_lt_one) h
+  · simp_all
+    simp_rw [card_insert_of_not_mem h1]
+    exact Or.resolve_left (le_iff_eq_or_lt.mp h.1) h.2
+
+theorem T_map_invol (t : Finset σ × σ) (h : t ∈ pairs σ k) :
+    T_map_restr σ (T_map_restr σ t h) (T_map_pair σ t h) = t := by
+  simp_rw [T_map_restr, T_map]
+  split_ifs with h1 h2 h3
+  · simp at h2
+  · simp_rw [insert_erase h1]
+  · simp_rw [erase_eq_self.mpr h1]
+    simp_all
+  · simp at h3
+
+open Classical in
+theorem weight_compose_T (t : Finset σ × σ) (h : t ∈ pairs σ k) :
+    (weight σ R k t) + weight σ R k (T_map_restr σ t h) = 0 := by
+  simp_rw [T_map_restr, T_map, weight]
+  simp_rw [pairs, mem_filter, PairsPred] at h
+  have h2 (n : ℕ) : -(-1 : MvPolynomial σ R) ^ n = (-1) ^ (n + 1)
+  · rw [← neg_one_mul ((-1 : MvPolynomial σ R) ^ n), pow_add, pow_one, mul_comm]
+  split_ifs with h1
+  · simp_rw [card_erase_of_mem h1, ← prod_erase_mul t.fst (fun j ↦ (X j : MvPolynomial σ R)) h1,
+      mul_comm, mul_assoc (∏ a in erase t.fst t.snd, X a), ← mul_add]
+    nth_rewrite 1 [← pow_one (X t.snd)]
+    simp_rw [← pow_add, add_comm]
+    have h3 : card t.fst ≥ 1
+    · have h4 : t.fst.Nonempty
+      · use t.snd
+        apply h1
+      exact lt_iff_add_one_le.mp (card_pos.mpr h4)
+    rw [← tsub_tsub_assoc h.right.left h3,
+      ← neg_neg ((-1 : MvPolynomial σ R) ^ (card t.fst - 1)), h2 (card t.fst - 1),
+      Nat.sub_add_cancel]
+    simp
+    exact h3
+  · simp_rw [card_insert_of_not_mem h1, prod_insert h1, mul_comm, mul_assoc (∏ a in t.fst, X a),
+      ← mul_add]
+    nth_rewrite 2 [← pow_one (X t.snd)]
+    have h3 : card t.fst + 1 ≤ k
+    · have ht1 := h.right.right
+      contrapose! ht1
+      exact And.intro (le_antisymm h.right.left (le_of_lt_succ ht1)) h1
+    simp_rw [← pow_add, ← Nat.add_sub_assoc h3, add_comm, add_tsub_add_eq_tsub_right]
+    rw [← neg_neg ((-1 : MvPolynomial σ R) ^ (card t.fst)), h2]
+    simp
+
+theorem weight_zero_for_fixed_by_T (t : Finset σ × σ) (h : t ∈ pairs σ k)
```

Comment:
The name seems to make more sense for the contraposition of this statement. It might be simpler that way as well:

```lean
theorem weight_zero_for_fixed_by_T' (t : Finset σ × σ) (h : t ∈ pairs σ k)
    (h2 : T_map_restr σ t h = t) : weight σ R k t = 0 := by
  have h3 := weight_compose_T σ R t h
  rw [h2, ← two_mul, _root_.mul_eq_zero] at h3
  exact h3.resolve_left two_ne_zero

theorem weight_zero_for_fixed_by_T (t : Finset σ × σ) (h : t ∈ pairs σ k)
    (h1 : weight σ R k t ≠ 0) : T_map_restr σ t h ≠ t :=
  mt (weight_zero_for_fixed_by_T' σ R t h) h1
```

Reply:
Done!